src/bio/bio_buffer.c

```diff
@@ -510,6 +515,69 @@ chunk_get_idle(struct bio_dma_buffer *bdb, struct bio_dma_chunk **chk_ptr)
 	return 0;
 }
 
+static unsigned int
+consumed_pgs(struct bio_dma_buffer *bdb)
+{
+	return (bdb->bdb_tot_cnt * bio_chk_sz) + bdb->bdb_huge_pgs;
+}
+
+static int
+chunk_get_huge(struct bio_dma_buffer *bdb, unsigned int pg_cnt,
+	       struct bio_dma_chunk **chk_ptr)
+{
+	struct bio_dma_chunk	*chk;
+	unsigned int		 tot_pgs = bio_chk_cnt_max * bio_chk_sz;
+	unsigned int		 reclaim_chks, reclaim_pgs;
+	int			 rc;
+
+	D_ASSERT(bio_chk_sz > 0);
+	/* Supported maximum IOV size is 1/2 DMA buffer upper bound size */
+	if (pg_cnt >= (tot_pgs / 2))
+		return -DER_REC2BIG;
+
+	/* Inflight huge IOV can consume at most 2/3 DMA buffer upper bound */
+	if (bdb->bdb_huge_pgs >= (tot_pgs * 2 / 3))
+		return -DER_AGAIN;
+
+	reclaim_chks = ((pg_cnt + bio_chk_sz - 1) / bio_chk_sz) * 2;
+	reclaim_pgs = reclaim_chks * bio_chk_sz;
+
+	/* Try to reclaim some idle chunks first */
+	if ((consumed_pgs(bdb) + reclaim_pgs) > tot_pgs)
+		dma_buffer_shrink(bdb, reclaim_chks);
+
+	/* Try to reclaim unused bulk cache */
+	if ((consumed_pgs(bdb) + reclaim_pgs) > tot_pgs) {
+		unsigned int	reclaimed = 0;
+
+		while (reclaimed < reclaim_chks) {
+			rc = bulk_reclaim_chunk(bdb, NULL);
+			if (rc)
+				break;
+			reclaimed++;
+		}
+		dma_buffer_shrink(bdb, reclaimed);
+	}
+
+	/* Not enough free buffer, retry later when other inflight I/O complete */
+	if ((consumed_pgs(bdb) + pg_cnt) > tot_pgs)
+		return -DER_AGAIN;
+
+	chk = dma_alloc_chunk(pg_cnt);
+	if (chk != NULL) {
+		chk->bdc_pg_idx = pg_cnt;
+		rc = 0;
+	} else if ((consumed_pgs(bdb) + reclaim_pgs) > tot_pgs) {
+		/* Retry if allocation failed when free buffer is less then (2 * required) */
```

Comment:
s/less then/less than/

So that means either the free buffer count is less than the required count or the left ones are not continuous to match the requirement, right? Return "-DER_NOMEM" means no retry later, right? Then why (2 * required) is the line between "-DER_AGAIN" and "-DER_NOMEM"?

Reply:
Right, -DER_NOMEM means no retry. In theory, allocation should success when available buffer is larger than required size, however, DPDK hugepage allocation could fail due to fragmentation or the space overhead from allocator when the free memory approaching zero, so I choose to return -DER_AGAIN when available buffer is less than "2 * required", so that it can retry later when more memory released by other inflight I/Os.